runtime/derive/src/lib.rs

```diff
@@ -65,18 +179,68 @@ fn asc_type_derive_struct(item_struct: ItemStruct) -> TokenStream {
         Fields::Named(fields) => fields.named.iter().map(|field| &field.ty),
         _ => panic!("AscType can only be derived for structs with named fields"),
     };
+    let field_types2 = field_types.clone();
+
+    // if no fields, return immediately
+    let (no_fields_return_bytes, no_fields_return_self) = if field_names.is_empty() {
+        (
+            quote! {
+                #![allow(unreachable_code)]
+                return Ok(Vec::new());
+            },
+            quote! {
+                #![allow(unreachable_code)]
+                return Ok(Self {});
+            },
+        )
+    } else {
+        (quote! {}, quote! {})
+    };
 
     TokenStream::from(quote! {
         impl #impl_generics graph::runtime::AscType for #struct_name #ty_generics #where_clause {
             fn to_asc_bytes(&self) -> Result<Vec<u8>, graph::runtime::DeterministicHostError> {
+                #no_fields_return_bytes
+
                 let in_memory_byte_count = std::mem::size_of::<Self>();
                 let mut bytes = Vec::with_capacity(in_memory_byte_count);
-                #(bytes.extend_from_slice(&self.#field_names.to_asc_bytes()?);)*
 
-                // Right now we do not properly implement the alignment rules dicted by
-                // AssemblyScript. As such, we here enforce that the structure is tighly
-                // packed and that no implicit padding has been added.
-                //
+                let mut offset = 0;
+                // max field alignment will also be struct alignment which we need to pad the end
+                let mut max_align = 0;
+
+                #(
+                    let field_align = std::mem::align_of::<#field_types>();
```

Comment:
I'm a bit worried by this because this uses Rust rules to determine the alignment of the field and not AssemblyScript rules.

You would think they are fully equivalent but usage of `std::mem::size_of` above tends to show that it's untrue that those methods are fully equivalent to AssemblyScript rules in all occasions.

I did a quick test however based on https://github.com/AssemblyScript/assemblyscript/blob/67457f3824efe80cfcb34465bd901c10c2e0c15c/tests/compiler/builtins.ts#L424-L428 (seems to be log2 values) and all usage for base types seems to work properly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0ad7be7dfdafad100557b4fe41bcd80d.

Anyway, hopefully it works for every cases that we require, but that is not 100% sure to me.